Subject: Linux-RT 2.6.25-rt
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
From ghaskins@novell.com Mon Mar 24 17:47:08 2008
Date: Mon, 10 Mar 2008 12:20:57 -0400
From: Gregory Haskins <ghaskins@novell.com>
To: rostedt@goodmis.org, mingo@elte.hu, tglx@linutronix.de
Cc: linux-rt-users@vger.kernel.org, ghaskins@novell.com
Subject: [PATCH] keep rd->online and cpu_online_map in sync

    [ The following text is in the "utf-8" character set. ]
    [ Your display is set for the "iso-8859-1" character set.  ]
    [ Some characters may be displayed incorrectly. ]

Hi Ingo, Steve, Thomas,
  Ingo pointed me at an issue on LKML/mainline regarding cpu-hotplug.
It turns out that the issue (I believe) was related to the root-domain
code that I added a few months ago.  This code is also in -rt, so the
bug should exist there as well.  This may fix the root-cause of that
s2ram bug that Mike Galbraith found and patched a while back as well.
If that is the case, we can probably get rid of the "num_online_cpus()"
in the update_migration code.  That will require confirmation from
Mike, however, as I do not have a P4 machine like his that exhibits
the bug.  For now, both patches should probably co-exist together.

This applies to 24.3-rt3

Regards,
-Greg 

------------------------------
keep rd->online and cpu_online_map in sync

It is possible to allow the root-domain cache of online cpus to
become out of sync with the global cpu_online_map.  This is because we
currently trigger removal of cpus too early in the notifier chain.
Other DOWN_PREPARE handlers may in fact run and reconfigure the
root-domain topology, thereby stomping on our own offline handling.

The end result is that rd->online may become out of sync with
cpu_online_map, which results in potential task misrouting.

So change the offline handling to be more tightly coupled with the
global offline process by triggering on CPU_DYING intead of
CPU_DOWN_PREPARE.

Signed-off-by: Gregory Haskins <ghaskins@novell.com>
---

 kernel/sched.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: linux-2.6.24.4-rt4/kernel/sched.c
===================================================================
--- linux-2.6.24.4-rt4.orig/kernel/sched.c	2008-03-24 19:07:17.000000000 -0400
+++ linux-2.6.24.4-rt4/kernel/sched.c	2008-03-24 19:07:19.000000000 -0400
@@ -6076,7 +6076,7 @@ migration_call(struct notifier_block *nf
 		spin_unlock_irq(&rq->lock);
 		break;
 
-	case CPU_DOWN_PREPARE:
+	case CPU_DYING:
 		/* Update our root-domain */
 		rq = cpu_rq(cpu);
 		spin_lock_irqsave(&rq->lock, flags);
